apps/dav/lib/Tree.php

```
@@ -0,0 +1,157 @@
+		if (strpos(rtrim($path, '/'), 'files/') !== 0) {
+			return parent::getNodeForPath($path);
+		}
+
+		// anything else will go through this shortcut to file nodes to avoid
+		// traversing every parent because it would trigger
+		// additional filecache queries and also
+		// additional locking of parent nodes and
+		// potential rescan in the case of external storages with update detection
+		$sections = explode('/', $path);
+		array_shift($sections);
+		$userId = array_shift($sections);
+
+		// this will ensure that the user exists and is accessible and also
+		// use the cached version when needed
+		$filesRoot = parent::getNodeForPath('files/' . $userId);
+
+		$fileView = new View('/' . $userId . '/files/');
+		$path = implode('/', $sections);
+
+		$path = trim($path, '/');
+
+		// is it the root node?
+		if ($path === '') {
+			return $filesRoot;
+		}
+
+		if (isset($this->fileNodeCache[$path]) && $this->fileNodeCache[$path] !== false) {
+			return $this->fileNodeCache[$path];
+		}
+
+		// check the path, also called when the path has been entered manually eg via a file explorer
+		if (\OC\Files\Filesystem::isForbiddenFileOrDir($path)) {
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
+
+		if ($path) {
+			try {
+				$fileView->verifyPath($path, basename($path));
+			} catch (\OCP\Files\InvalidPathException $ex) {
+				throw new InvalidPath($ex->getMessage());
+			}
+		}
+
+		// read from file cache
+		try {
+			$info = $fileView->getFileInfo($path);
+		} catch (StorageNotAvailableException $e) {
+			throw new \Sabre\DAV\Exception\ServiceUnavailable('Storage is temporarily not available', 0, $e);
+		} catch (StorageInvalidException $e) {
+			throw new \Sabre\DAV\Exception\NotFound('Storage ' . $path . ' is invalid');
+		} catch (LockedException $e) {
+			throw new \Sabre\DAV\Exception\Locked();
+		} catch (ForbiddenException $e) {
+			throw new \Sabre\DAV\Exception\Forbidden();
+		}
+
+		if (!$info) {
+			$this->fileNodeCache[$cachePath] = false;
```

I'd rather have a clear expectation about missing fileInfos:
* If the expectation is NOT to cache the failure and let it retry afterwards, this isn't needed. In addition, the `$cachePath` variable isn't defined.
* If the expectation is to cache the failure so next call will return the same error, we should return the cached error as soon as possible.

If the valid results are being cached, I expect the errors are cached too. If this won't be the case, the PHPDoc should state that.

---

yes, now throwing an exception if we detect a cached `false`